### DISCUSSION TITLE
how much is reliable the message delivery in gossipsub

### DISCUSSION THREAD

**Discussion by revoltez** [libp2p/rust-libp2p#3717] [Q&A]
How does the libp2p GossipSub protocol handle message delivery in the event of a network partition, specifically with respect to messages that were lost during the partition? are lost messages retrieved upon network re-connection automatically by the protocol or are they permanently dropped and must be re-gossiped or requested manually? in other words, does the protocol provide a reliable delivery mechanism even after node crashes, or is it only a best-effort protocol?
another side question is about newly joined peers, will they receive old topic messages that they subscribed to or just receive the new ones?

**Comment by dariusc93** [comment]
To quickly answer the side question: peers joining a topic would not be able to see older messages, only new ones so long as they are subscribed (not including other factors such as connectivity, etc). For the peer to receive older messages, that would be up to the implementation of libp2p to decide, such as sending a request from peer(s) on the topic for older messages, having a rendezvous that would store previous messages, etc. This is mostly up to you to decide, though this would require a little more work from your end if you wish for the peer(s) to receive older messages.

**Reply by mxinden** [reply]
To add to the above, I suggest using gossipsub in the following way:

- Gossip the deltas of the current state around, i.e. only gossip new changes
- Have new (or previously crashed) nodes listen in on deltas.
- In case they notice that their local state is behind, have an additional reconciliation protocol, enabling them to catch up to the current state.
- Once caught up, have such node stay in sync with the gossip messages from step (1).

**Reply by revoltez** [reply]
@mxinden @dariusc93  thanks for the reply, so basically gossipsub should be considered as best effort delivery (trying to be as reliable as it can with no gurantees)

**Reply by mxinden** [reply]
You will need some kind of reliable synchronization mechanism anyways to catch up new nodes joining the network. Might as well use that same mechanism to catch up nodes that have crashed recently.

**Reply by mxinden** [reply]
You might want to look into https://github.com/ipfs-rust/libp2p-bitswap to synchronize larger chunks of state.
